core/src/Cuda/Kokkos_Cuda_BlockSize_Deduction.hpp

```diff
@@ -25,17 +25,60 @@
 namespace Kokkos {
 namespace Impl {
 
+inline int cuda_warp_per_sm_allocation_granularity(cudaDeviceProp const& properties) {
+  // Allocation granularity of warps in each sm
+  switch (properties.major) {
+    case 3:
+    case 5:
+    case 7:
+    case 8:
+    case 9: return 4;
+    case 6: return (properties.minor == 0 ? 2 : 4);
+    default:
+      throw_runtime_exception("Unknown device in cuda warp per sm allocation granularity");
+      return 0;
+  }
+}
+
+inline int cuda_max_warps_per_sm_registers(cudaDeviceProp const& properties,
+                                           cudaFuncAttributes const& attributes) {
+  // Maximum number of warps per sm as a function of register counts,
+  // subject to the constraint that warps are allocated with a fixed granularity
+  int const max_regs_per_block = properties.regsPerBlock;
+  int const regs_per_warp = attributes.numRegs * properties.warpSize;
+  int const warp_granularity = cuda_warp_per_sm_allocation_granularity(properties);
+  // The granularity of register allocation is chunks of 256 registers per warp,
+  // which implies a need to over-allocate, so we round up
+  int const allocated_regs_per_warp = (regs_per_warp + 256 - 1) / 256;
+
+  // The maximum number of warps per SM is constrained from above by register allocation.
+  // To satisfy the constraint that warps per SM is allocated at a finite granularity,
+  // we need to round down.
+  int const max_warps_per_sm = warp_granularity * (max_regs_per_block / (allocated_regs_per_warp * warp_granularity));
```

Comment:
Does this work if you use `-ffast-math` ?

Reply:
I haven't explicitly tested it (yet), but based on some quick reading online I don't think `-ffast-math` touches integer division? (If it did, I feel like it would break a handful of other codes...)

Reply:
I am not sure that's why I ask. With floating-point this wouldn't work but I don't know about integer

Reply:
Just tested it in godbolt, it generates `imul`, `idiv`, `imul` whether or not `-ffast-math` is set: https://godbolt.org/z/G3sz4hjfq

Reply:
The standard defines integer arithmetic semantics much more concretely than floating point. Overflow is UB, everything else has a precise result.